tap_salesforce/client.py

```diff
@@ -255,11 +257,14 @@ def _make_request(self, method, path, data=None, params=None) -> requests.Respon
             method, url, headers=headers, params=params, data=data
         )
 
-        json = resp.json()
-        if isinstance(json, list) and isinstance(json[0], dict) and json[0]['errorCode'] == 'QUERY_TIMEOUT':
-            raise QueryTimeoutError()
-
-        resp.raise_for_status()
+        if resp.status_code < 200 or resp.status_code > 299:
+            # Salesforce error json body looks like
+            # [{'message': 'Your query request was running for too long.', 'errorCode': 'QUERY_TIMEOUT'}]
+            err_array = resp.json()
+            if not isinstance(err_array, list) or not isinstance(err_array[0], dict):
+                resp.raise_for_status()
+            
+            raise SalesforceException(err_array[0]['message'], err_array[0]['errorCode'])
```

Comment:
It would be really neat, if we changed the `SalesforceException.__init__` function to simply accept the error array. That way, we can centralize the knowledge of "how does a Salesforce error look like" there.

```python
class SalesforceException(Exception):
    def __init__(self, errors):
        self.errors = errors

        if len(self.errors) == 0:
            super().__init__("an unknown error occurred")
        elif len(self.errors) == 1:
            super().__init__(f"{self.errors[0]['message']}")
        else:
            super().__init__(f"{self.errors[0]['message']} (+{len(self.errors)-1} more)")

raise SalesforceException(err_array)
```

Reply:
check https://github.com/dreamdata-io/tap-salesforce/commit/9dcdc2c26d2607e6224afe5da862a41a0fe6d242 - which tries to centralize the error to the exceptions.py. It has the same idea as your suggestion but implmented a bit differently. Do let me know if https://github.com/dreamdata-io/tap-salesforce/commit/9dcdc2c26d2607e6224afe5da862a41a0fe6d242 is good enough